src/Components/src/Microsoft.AspNetCore.Components/Rendering/HtmlRenderer.cs

```diff
@@ -0,0 +1,234 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.RenderTree;
+
+namespace Microsoft.AspNetCore.Components.Rendering
+{
+    /// <summary>
+    /// A <see cref="Renderer"/> that produces HTML.
+    /// </summary>
+    public class HtmlRenderer : Renderer
+    {
+        private static readonly HashSet<string> SelfClosingElements = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "area", "base", "br", "col", "embed", "hr", "img", "input", "link", "meta", "param", "source", "track", "wbr"
+        };
+
+        private readonly HtmlEncoder _htmlEncoder;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="HtmlRenderer"/>.
+        /// </summary>
+        /// <param name="serviceProvider">The <see cref="IServiceProvider"/> to use to instantiate components.</param>
+        /// <param name="htmlEncoder">The <see cref="HtmlEncoder"/> to use for encoding content.</param>
+        public HtmlRenderer(IServiceProvider serviceProvider, HtmlEncoder htmlEncoder) : base(serviceProvider)
+        {
+            _htmlEncoder = htmlEncoder;
+        }
+
+        /// <inheritdoc />
+        protected override Task UpdateDisplayAsync(in RenderBatch renderBatch)
+        {
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Renders a component into a sequence of <see cref="string"/> fragments that represent the textual representation
+        /// of the HTML produced by the component.
+        /// </summary>
+        /// <typeparam name="T">The type of the <see cref="IComponent"/>.</typeparam>
+        /// <param name="initialParameters">A <see cref="ParameterCollection"/> with the initial parameters to render the component.</param>
+        /// <returns>A sequence of <see cref="string"/> fragments that represent the HTML text of the component.</returns>
+        public IEnumerable<string> RenderComponent<T>(ParameterCollection initialParameters) where T : IComponent
+        {
+            return RenderComponent(typeof(T), initialParameters);
+        }
+
+        /// <summary>
+        /// Renders a component into a sequence of <see cref="string"/> fragments that represent the textual representation
+        /// of the HTML produced by the component.
+        /// </summary>
+        /// <param name="componentType">The type of the <see cref="IComponent"/>.</param>
+        /// <param name="initialParameters">A <see cref="ParameterCollection"/> with the initial parameters to render the component.</param>
+        /// <returns>A sequence of <see cref="string"/> fragments that represent the HTML text of the component.</returns>
+        private IEnumerable<string> RenderComponent(Type componentType, ParameterCollection initialParameters)
+        {
+            var result = new List<string>();
```

Comment:
Minor nitpick, but this could be moved into the block starting on L72, so it's not used if `frames.Count == 0`.